src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -294,12 +298,31 @@ public async Task HardDeleteAsync(ResourceKey key, CancellationToken cancellatio
 
         public async Task UpdateSearchParameterIndicesBatchAsync(IReadOnlyCollection<ResourceWrapper> resources, CancellationToken cancellationToken)
         {
-            // TODO: use batch command to update both hash values and search index values for list updateSearchIndices
-
-            // this is a place holder update until we batch update resources
-            foreach (var resource in resources)
+            using (SqlConnectionWrapper sqlConnectionWrapper = await _sqlConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken, true))
+            using (SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand())
             {
-                await UpdateSearchIndexForResourceAsync(resource, WeakETag.FromVersionId(resource.Version), cancellationToken);
+                VLatest.BulkReindexResources.PopulateCommand(
+                    sqlCommandWrapper,
+                    _bulkReindexResourcesTvpGeneratorVLatest.Generate(resources.ToList()));
+
+                try
+                {
+                    await sqlCommandWrapper.ExecuteNonQueryAsync(cancellationToken);
+                }
+                catch (SqlException e)
+                {
+                    switch (e.Number)
+                    {
+                        // TODO: we should attempt to reindex resources that failed to be reindexed
+                        case SqlErrorCodes.PreconditionFailed:
+                            throw new PreconditionFailedException(string.Format(Core.Resources.ReindexingResourceVersionConflict));
+                        case SqlErrorCodes.NotFound:
+                            throw new ResourceNotFoundException(string.Format(Core.Resources.ReindexingResourceNotFound));
+                        default:
```

Comment:
You could probably also end up with deadlocking as well

Reply:
I can see how the `UPDLOCK` on the index `IX_Resource_ResourceTypeId_ResourceId` could block other transactions, and I can see transactions being rolled back because we have RCSI enabled - but what is the second resource that would lead to a deadlock?

Reply:
As discussed, it makes sense to add retry logic at this layer for both precondition failed exceptions and exceptions generated by SQL Server. 

Taking a second look at this, I'll plan to keep it as-is for now, since the next step seems to be to add the retry logic. 

I updated these two stories to reflect what we chatted about:

[AB#79925](https://microsofthealth.visualstudio.com/Health/_workitems/edit/79925) (which is SQL specific)
[AB#80638](https://microsofthealth.visualstudio.com/Health/_workitems/edit/80638) (which is for both Cosmos and SQL)